### ISSUE TITLE
False positive error with `no-param-reassign` rule

### ISSUE THREAD

**Issue by yekver** [eslint/eslint#8670]
In this case copied object reference will be reassigned, but no mutation will be done.

* **ESLint Version: v3.19.0**
* **Node Version: v7.10.0**
* **npm Version: v4.2.0**

**What parser (default, Babel-ESLint, etc.) are you using?**
default, "ecmaVersion": 6

```
var list = [{a: 1}, {b: 2}, {c: 3}];
list.reduce((res, obj, i) => {
    obj = {k: 7};
    res[i]=obj;

    return res;
}, []);
```

**What did you expect to happen?**

no error should be provided

**What actually happened? Please include the actual, raw output from ESLint.**

`error  Assignment to function parameter 'obj' no-param-reassign`

<img width="1154" alt="2017-05-31 12 38 45" src="https://cloud.githubusercontent.com/assets/1935877/26627044/bc2373d6-4601-11e7-9271-266d1394db6e.png">

**Comment by not-an-aardvark** [comment]
This is working as intended. The rule reports when a function parameter is reassigned, even though it's not mutated.

**Comment by not-an-aardvark** [comment]
Closing because this is working as intended.
